src/compiler/scala/tools/nsc/ast/TreeGen.scala

```diff
@@ -156,6 +155,9 @@ abstract class TreeGen extends scala.reflect.internal.TreeGen with TreeDSL {
   // let's assume for now annotations don't affect casts, drop them there, and bring them back using the outer Typed tree
   def mkCastPreservingAnnotations(tree: Tree, pt: Type) =
     Typed(mkCast(tree, pt.withoutAnnotations.dealias), TypeTree(pt))
+    // ^^^ I think we should either normalize or do nothing, but the half measure of dealias does not make sense,
+    // as the logic behind a cast operates on the fully normalized type, not just on a dealiased type (think refinements with type aliases).
+    // It would be ok to do nothing here, because erasure will convert the type to something that can be casted anyway.
```

Comment:
I was in the neighbourhood in https://github.com/scala/scala/pull/5722/files#diff-36cc56dc4b99ae4c41d8f122c5e9dce9R150 and the sky didn't fall when I added auto-dealiasing.

Reply:
I'll follow up with cleanup in 2.13.x